Build failure: perlPackages.CryptDES

## Steps To Reproduce

Steps to reproduce the behavior:

1. build `perlPackages.CryptDES`

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/gl66pqmbkfvy5mxsz02dm0dgr2ikzzim-Crypt-DES-2.07.tar.gz
source root is Crypt-DES-2.07
setting SOURCE_DATE_EPOCH to timestamp 1374612172 of file Crypt-DES-2.07/META.yml
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
patching ./Makefile.PL...
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for Crypt::DES
Writing MYMETA.yml and MYMETA.json
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
build flags: SHELL=/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash
cp DES.pm blib/lib/Crypt/DES.pm
Running Mkbootstrap for DES ()
chmod 644 "DES.bs"
"/nix/store/1i003ijlh9i0mzp6alqby5hg3090pjdx-perl-5.40.0/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- DES.bs blib/arch/auto/Crypt/DES/DES.bs 644
"/nix/store/1i003ijlh9i0mzp6alqby5hg3090pjdx-perl-5.40.0/bin/perl" "/nix/store/1i003ijlh9i0mzp6alqby5hg3090pjdx-perl-5.40.0/lib/perl5/5.40.0/ExtUtils/xsubpp" >
mv DES.xsc DES.c
cc -c   -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/no-such-path/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS>
DES.xs: In function 'XS_Crypt__DES_expand_key':
DES.xs:39:17: error: implicit declaration of function 'perl_des_expand_key'; did you mean '_des_expand_key'? [-Wimplicit-function-declaration]
   39 |                 perl_des_expand_key((i8 *)key, ks);
      |                 ^~~~~~~~~~~~~~~~~~~
      |                 _des_expand_key
DES.xs: In function 'XS_Crypt__DES_crypt':
DES.xs:69:17: error: implicit declaration of function 'perl_des_crypt'; did you mean '_des_crypt'? [-Wimplicit-function-declaration]
   69 |                 perl_des_crypt(input, SvGROW(output, output_len), (i32 *)ks, enc_flag);
      |                 ^~~~~~~~~~~~~~
      |                 _des_crypt
make: *** [Makefile:335: DES.o] Error 1
```

</details>

## Additional context

<!-- Add any other context about the problem here. -->

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.7, NixOS, 25.05 (Warbler), 25.05.20241231.93097ab`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.92.0-dev-pre20241230-5c7ea4f
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/anton/.config/nix/nix.conf:/nix/store/zs83w398qva6syb97mb18i7mbas33a89-stylix-kde-config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/anton/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/anton/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/anton/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/8gqhnjl4l26ksa02arw7k5axiip9by9b-lix-2.92.0-dev-pre20241230-5c7ea4f/share`
 - channels(root): `""`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`


## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
There's no maintainer for this package.

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/perl-modules/CryptDES-gcc14.patch
+++ b/pkgs/development/perl-modules/CryptDES-gcc14.patch
@@ -1,0 +1,8 @@
+--- a/_des.h
++++ b/_des.h
+@@ -4,4 +4,6 @@
+ 
+ void _des_crypt( des_cblock in, des_cblock out, des_ks key, int encrypt );
+ void _des_expand_key( des_user_key userKey, des_ks key );
++void perl_des_expand_key( des_user_key userKey, des_ks key );
++void perl_des_crypt( des_cblock input, des_cblock output, des_ks ks, int encrypt );

--- a/pkgs/top-level/perl-packages.nix
+++ b/pkgs/top-level/perl-packages.nix
@@ -5018,6 +5018,10 @@ with self; {
       url = "mirror://cpan/authors/id/D/DP/DPARIS/Crypt-DES-2.07.tar.gz";
       hash = "sha256-LbHrtYN7TLIAUcDuW3M7RFPjE33wqSMGA0yGdiHt1+c=";
     };
+    patches = [
+      # add extra definitions in header to please -Werror=implicit-function-declaration
+      ../development/perl-modules/CryptDES-gcc14.patch
+    ];
     meta = {
       description = "Perl DES encryption module";
       license = with lib.licenses; [ bsdOriginalShortened ];

